File: workers/python/python/tasker_core/registry/resolver_chain.py
"""TAS-93: Resolver Chain - Priority-Ordered Handler Resolution.

The ResolverChain orchestrates handler resolution by trying resolvers
in priority order until one successfully resolves the handler.

Resolution Contract:
1. If HandlerDefinition has `resolver` hint, use ONLY that resolver
2. Otherwise, try resolvers in priority order (lower = first)
3. Wrap resolved handler with MethodDispatchWrapper if needed
4. Return None if no resolver can handle the definition

Default Chain (when using .default()):
- Priority 10:  ExplicitMappingResolver - registered handlers
- Priority 100: ClassLookupResolver     - class path inference

Usage:
    # Create default chain
    chain = ResolverChain.default()

    # Or build custom chain
    chain = ResolverChain()
    chain.add_resolver(MyResolver())
    chain.add_resolver(AnotherResolver())

    # Resolve a handler
    handler = chain.resolve(definition, config)

Adding Custom Resolvers:
    chain.add_resolver(CustomResolver())

Named Resolver Access:
    # Definition with resolver hint
    definition = HandlerDefinition(callable="x", resolver="payment_resolver")
    # Chain will ONLY use PaymentResolver
"""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING, Any

from ..logging import log_debug, log_warn
from .method_dispatch_wrapper import MethodDispatchWrapper

if TYPE_CHECKING:
    from .base_resolver import BaseResolver
    from .handler_definition import HandlerDefinition


class ResolverNotFoundError(Exception):
    """Error raised when resolver hint points to unknown resolver."""

    pass


class ResolutionError(Exception):
    """Error raised when resolution fails completely."""

    pass


class ResolverChain:
    """Priority-ordered chain of handler resolvers.

    Orchestrates handler resolution by trying resolvers in priority order.
    Supports resolver hints to bypass the chain and use a specific resolver.

    Attributes:
        resolvers: List of resolvers in priority order.
        resolvers_by_name: Mapping of resolver names to resolvers.
    """

    def __init__(self) -> None:
        """Initialize an empty resolver chain."""
        self._resolvers: list[BaseResolver] = []
        self._resolvers_by_name: dict[str, BaseResolver] = {}
        self._lock = threading.RLock()

    @classmethod
    def default(cls) -> ResolverChain:
        """Create a chain with default resolvers.

        Returns:
            Chain with ExplicitMapping + ClassLookup resolvers.
        """
        from .resolvers import ClassLookupResolver, ExplicitMappingResolver

        chain = cls()
        chain.add_resolver(ExplicitMappingResolver())
        chain.add_resolver(ClassLookupResolver())
        return chain

    def add_resolver(self, resolver: BaseResolver) -> ResolverChain:
        """Add a resolver to the chain.

        Resolvers are automatically sorted by priority (lower = first).

        Args:
            resolver: Resolver to add.

        Returns:
            Self for method chaining.
        """
        with self._lock:
            self._resolvers.append(resolver)
            self._resolvers.sort(key=lambda r: r.priority)
            self._resolvers_by_name[resolver.name] = resolver
        return self

    def remove_resolver(self, name: str) -> BaseResolver | None:
        """Remove a resolver by name.

        Args:
            name: Resolver name to remove.

        Returns:
            Removed resolver or None if not found.
        """
        with self._lock:
            resolver = self._resolvers_by_name.pop(name, None)
            if resolver:
                self._resolvers.remove(resolver)
            return resolver

    def get_resolver(self, name: str) -> BaseResolver | None:
        """Get a resolver by name.

        Args:
            name: Resolver name.

        Returns:
            Resolver or None if not found.
        """
        return self._resolvers_by_name.get(name)

    @property
    def explicit_resolver(self) -> BaseResolver | None:
        """Get the explicit mapping resolver (convenience method).

        Returns:
            ExplicitMappingResolver or None.
        """
        return self._resolvers_by_name.get("explicit_mapping")

    def resolve(
        self,
        definition: HandlerDefinition,
        config: dict[str, Any] | None = None,
    ) -> Any | None:
        """Resolve a handler from the definition.

        Args:
            definition: Handler configuration.
            config: Additional context.

        Returns:
            Resolved handler or None.

        Raises:
            ResolverNotFoundError: If resolver hint points to unknown resolver.
        """
        config = config or {}

        # Contract: If resolver hint is present, use ONLY that resolver
        if definition.has_resolver_hint():
            return self._resolve_with_hint(definition, config)

        # Otherwise: Try inferential chain resolution
        return self._resolve_with_chain(definition, config)

    def can_resolve(
        self,
        definition: HandlerDefinition,
        config: dict[str, Any] | None = None,
    ) -> bool:
        """Check if any resolver can handle this definition.

        Args:
            definition: Handler configuration.
            config: Additional context.

        Returns:
            True if definition can be resolved.
        """
        config = config or {}

        if definition.has_resolver_hint():
            resolver = self._resolvers_by_name.get(definition.resolver or "")
            return resolver.can_resolve(definition, config) if resolver else False

        return any(r.can_resolve(definition, config) for r in self._resolvers)

    def registered_callables(self) -> list[str]:
        """Get all registered callables across all resolvers.

        Returns:
            List of all registered callable identifiers.
        """
        callables: list[str] = []
        for resolver in self._resolvers:
            callables.extend(resolver.registered_callables())
        return list(set(callables))

    def register(self, key: str, handler: Any) -> ResolverChain:
        """Register a handler directly (convenience method for ExplicitMappingResolver).

        Args:
            key: Handler key.
            handler: Handler class or instance.

        Returns:
            Self for method chaining.

        Raises:
            RuntimeError: If no ExplicitMappingResolver in chain.
        """
        explicit = self.explicit_resolver
        if explicit is None:
            raise RuntimeError("No ExplicitMappingResolver in chain")

        # Access the register method - we know it exists on ExplicitMappingResolver
        explicit.register(key, handler)  # type: ignore[attr-defined]
        return self

    def chain_info(self) -> list[dict[str, Any]]:
        """Get chain info for debugging.

        Returns:
            List of resolver info dicts.
        """
        return [
            {
                "name": resolver.name,
                "priority": resolver.priority,
                "callables": len(resolver.registered_callables()),
            }
            for resolver in self._resolvers
        ]

    def __len__(self) -> int:
        """Return number of resolvers in chain."""
        return len(self._resolvers)

    @property
    def resolver_names(self) -> list[str]:
        """Get names of resolvers in priority order.

        Returns:
            List of resolver names.
        """
        return [r.name for r in self._resolvers]

    def list_resolvers(self) -> list[tuple[str, int]]:
        """List resolvers with their priorities.

        Returns:
            List of (name, priority) tuples in priority order.
        """
        return [(r.name, r.priority) for r in self._resolvers]

    def _resolve_with_hint(
        self,
        definition: HandlerDefinition,
        config: dict[str, Any],
    ) -> Any | None:
        """Resolve using explicit resolver hint (bypasses chain).

        Args:
            definition: Handler configuration.
            config: Additional context.

        Returns:
            Resolved handler or None.

        Raises:
            ResolverNotFoundError: If resolver hint points to unknown resolver.
        """
        resolver_name = definition.resolver or ""
        resolver = self._resolvers_by_name.get(resolver_name)

        if resolver is None:
            log_warn(f"ResolverChain: Unknown resolver hint '{resolver_name}'")
            raise ResolverNotFoundError(f"Unknown resolver: '{resolver_name}'")

        handler = resolver.resolve(definition, config)

        if handler is None:
            log_warn(
                f"ResolverChain: Resolver '{resolver_name}' failed to resolve '{definition.callable}'"
            )
            return None

        log_debug(f"ResolverChain: Resolved '{definition.callable}' via hint '{resolver_name}'")
        return self.wrap_for_method_dispatch(handler, definition)

    def _resolve_with_chain(
        self,
        definition: HandlerDefinition,
        config: dict[str, Any],
    ) -> Any | None:
        """Resolve using chain (tries resolvers in priority order).

        Args:
            definition: Handler configuration.
            config: Additional context.

        Returns:
            Resolved handler or None.
        """
        for resolver in self._resolvers:
            if not resolver.can_resolve(definition, config):
                continue

            handler = resolver.resolve(definition, config)
            if handler is None:
                continue

            log_debug(f"ResolverChain: Resolved '{definition.callable}' via '{resolver.name}'")
            return self.wrap_for_method_dispatch(handler, definition)

        log_debug(f"ResolverChain: No resolver could handle '{definition.callable}'")
        return None

    def wrap_for_method_dispatch(
        self,
        handler: Any,
        definition: HandlerDefinition,
    ) -> Any:
        """Wrap handler for method dispatch if needed.

        Args:
            handler: Resolved handler.
            definition: Handler configuration.

        Returns:
            Handler (possibly wrapped).
        """
        # No wrapping needed if using default .call method
        if not definition.uses_method_dispatch():
            return handler

        effective_method = definition.effective_method()

        # Check if handler has the target method
        if not hasattr(handler, effective_method):
            log_warn(
                f"ResolverChain: Handler {handler.__class__.__name__} "
                f"doesn't have method '{effective_method}'"
            )
            return None

        log_debug(
            f"ResolverChain: Wrapping {handler.__class__.__name__} "
            f"for method dispatch → {effective_method}"
        )
        return MethodDispatchWrapper(handler, effective_method)

File: workers/ruby/lib/tasker_core/registry/resolver_chain.rb
# frozen_string_literal: true

require_relative 'resolvers/base_resolver'
require_relative 'resolvers/method_dispatch_wrapper'

module TaskerCore
  module Registry
    # TAS-93: Resolver Chain - Priority-Ordered Handler Resolution
    #
    # The ResolverChain orchestrates handler resolution by trying resolvers
    # in priority order until one successfully resolves the handler.
    #
    # == Resolution Contract
    #
    # 1. If HandlerDefinition has `resolver:` hint, use ONLY that resolver
    # 2. Otherwise, try resolvers in priority order (lower = first)
    # 3. Wrap resolved handler with MethodDispatchWrapper if needed
    # 4. Return nil if no resolver can handle the definition
    #
    # == Default Chain (when using .default)
    #
    # Priority 10:  ExplicitMappingResolver - registered handlers
    # Priority 100: ClassConstantResolver   - class path inference
    #
    # == Usage
    #
    #   # Create default chain
    #   chain = ResolverChain.default
    #
    #   # Or build custom chain
    #   chain = ResolverChain.new
    #     .with_resolver(MyResolver.new)
    #     .with_resolver(AnotherResolver.new)
    #
    #   # Resolve a handler
    #   handler = chain.resolve(definition, config)
    #
    # == Adding Custom Resolvers
    #
    #   chain.add_resolver(CustomResolver.new)
    #
    # == Named Resolver Access
    #
    #   # Add with name for resolver hint support
    #   chain.add_resolver(PaymentResolver.new)
    #
    #   # Definition with resolver hint
    #   definition.resolver = 'payment_resolver'
    #   # Chain will ONLY use PaymentResolver
    #
    class ResolverChain
      # Error raised when resolver hint points to unknown resolver
      class ResolverNotFoundError < StandardError; end

      # Error raised when resolution fails completely
      class ResolutionError < StandardError; end

      def initialize
        @resolvers = []
        @resolvers_by_name = {}
        @logger = TaskerCore::Logger.instance
      end

      # Create a chain with default resolvers
      #
      # @return [ResolverChain] Chain with Explicit + ClassConstant resolvers
      def self.default
        require_relative 'resolvers/explicit_mapping_resolver'
        require_relative 'resolvers/class_constant_resolver'

        new.tap do |chain|
          chain.add_resolver(Resolvers::ExplicitMappingResolver.new)
          chain.add_resolver(Resolvers::ClassConstantResolver.new)
        end
      end

      # Add a resolver to the chain
      #
      # @param resolver [Resolvers::BaseResolver] Resolver to add
      # @return [self] For method chaining
      def add_resolver(resolver)
        @resolvers << resolver
        @resolvers.sort_by!(&:priority)
        @resolvers_by_name[resolver.name] = resolver
        self
      end

      # Fluent API for adding resolvers
      alias with_resolver add_resolver

      # Remove a resolver by name
      #
      # @param name [String] Resolver name to remove
      # @return [Resolvers::BaseResolver, nil] Removed resolver
      def remove_resolver(name)
        resolver = @resolvers_by_name.delete(name)
        @resolvers.delete(resolver) if resolver
        resolver
      end

      # Get a resolver by name
      #
      # @param name [String] Resolver name
      # @return [Resolvers::BaseResolver, nil]
      def get_resolver(name)
        @resolvers_by_name[name]
      end

      # Get the explicit mapping resolver (convenience method)
      #
      # @return [Resolvers::ExplicitMappingResolver, nil]
      def explicit_resolver
        @resolvers_by_name['explicit_mapping']
      end

      # Resolve a handler from the definition
      #
      # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
      # @param config [Hash] Additional context
      # @return [Object, nil] Resolved handler or nil
      def resolve(definition, config = {})
        # Contract: If resolver hint is present, use ONLY that resolver
        return resolve_with_hint(definition, config) if definition.has_resolver_hint?

        # Otherwise: Try inferential chain resolution
        resolve_with_chain(definition, config)
      end

      # Check if any resolver can handle this definition
      #
      # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
      # @param config [Hash] Additional context
      # @return [Boolean]
      def can_resolve?(definition, config = {})
        if definition.has_resolver_hint?
          resolver = @resolvers_by_name[definition.resolver]
          return resolver&.can_resolve?(definition, config) || false
        end

        @resolvers.any? { |r| r.can_resolve?(definition, config) }
      end

      # Get all registered callables across all resolvers
      #
      # @return [Array<String>] List of all registered callable identifiers
      def registered_callables
        @resolvers.flat_map(&:registered_callables).uniq
      end

      # Register a handler directly (convenience method for ExplicitMappingResolver)
      #
      # @param key [String] Handler key
      # @param handler [Object] Handler class or instance
      # @return [self]
      def register(key, handler)
        explicit = explicit_resolver
        raise 'No ExplicitMappingResolver in chain' unless explicit

        explicit.register(key, handler)
        self
      end

      # Get chain info for debugging
      #
      # @return [Array<Hash>] Resolver info
      def chain_info
        @resolvers.map do |resolver|
          {
            name: resolver.name,
            priority: resolver.priority,
            callables: resolver.registered_callables.size
          }
        end
      end

      # @return [Integer] Number of resolvers in chain
      def size
        @resolvers.size
      end

      # @return [Array<String>] Names of resolvers in priority order
      def resolver_names
        @resolvers.map(&:name)
      end

      # Wrap handler for method dispatch if needed.
      #
      # This method is public because HandlerRegistry needs to wrap handlers
      # that it instantiates directly (outside the resolver chain flow).
      #
      # @param handler [Object] Resolved handler instance
      # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
      # @return [Object, nil] Handler (possibly wrapped), or nil if method not found
      def wrap_for_method_dispatch(handler, definition)
        # No wrapping needed if using default .call method
        return handler unless definition.uses_method_dispatch?

        effective_method = definition.effective_method.to_sym

        # Verify handler responds to the target method
        unless handler.respond_to?(effective_method)
          @logger.warn("ResolverChain: Handler #{handler.class} doesn't respond to ##{effective_method}")
          return nil
        end

        @logger.debug("ResolverChain: Wrapping #{handler.class} for method dispatch → ##{effective_method}")
        Resolvers::MethodDispatchWrapper.new(handler, effective_method)
      end

      private

      # Resolve using explicit resolver hint (bypasses chain)
      #
      # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
      # @param config [Hash] Additional context
      # @return [Object, nil] Resolved handler
      def resolve_with_hint(definition, config)
        resolver_name = definition.resolver
        resolver = @resolvers_by_name[resolver_name]

        unless resolver
          @logger.warn("ResolverChain: Unknown resolver hint '#{resolver_name}'")
          raise ResolverNotFoundError, "Unknown resolver: '#{resolver_name}'"
        end

        handler = resolver.resolve(definition, config)

        unless handler
          @logger.warn("ResolverChain: Resolver '#{resolver_name}' failed to resolve '#{definition.callable}'")
          return nil
        end

        @logger.debug("ResolverChain: Resolved '#{definition.callable}' via hint '#{resolver_name}'")
        wrap_for_method_dispatch(handler, definition)
      end

      # Resolve using chain (tries resolvers in priority order)
      #
      # @param definition [TaskerCore::Types::HandlerDefinition] Handler configuration
      # @param config [Hash] Additional context
      # @return [Object, nil] Resolved handler
      def resolve_with_chain(definition, config)
        @resolvers.each do |resolver|
          next unless resolver.can_resolve?(definition, config)

          handler = resolver.resolve(definition, config)
          next unless handler

          @logger.debug("ResolverChain: Resolved '#{definition.callable}' via '#{resolver.name}'")
          return wrap_for_method_dispatch(handler, definition)
        end

        @logger.debug("ResolverChain: No resolver could handle '#{definition.callable}'")
        nil
      end
    end
  end
end
